rfcs/proposed/flow_graph_fold_node/README.md

```diff
@@ -0,0 +1,226 @@
+computes partial folds for each element, it is unclear for the node when the computed result is final and all of the elements in the stream
+were processed. To handle this, it is required to take a special signal from the user indicating that no more elements from the input stream
+with the specified tag are expected. Some special type defined by the Flow Graph is required for that purpose, e.g. ``tbb::flow::fold_stream_end``.
+``fold_stream_end`` instance should be explicitly submitted to the node once the input stream end was reached:
+
+```cpp
+// Submit stream1
+for (auto item : stream1) {
+    f_node.try_put(input_type{stream1_index, item});
+}
+f_node.try_put(tbb::flow::fold_stream_end{stream1_index});
+
+for (auto item : stream2) {
+    f_node.try_put(input_type{stream2_index, item});
+}
+f_node.try_put(tbb::flow::fold_stream_end{stream2_index});
+
+int result = 0;
+buf.try_get(result);
+
+// result is expected to be either 25 (result of the first stream fold) or
+// 30 (result of the second stream fold)
+```
+
+From the implementation perspective, there can be multiple approaches for implementing the ``fold_stream_end`` that affects the user API.
+Some of them were considered in a [separate section](#fold_stream_end-implementation-approaches).
+
+## ``fold_stream_end`` Implementation Approaches
+
+Consider the possible implementation of ``fold_node`` described in this paper as a wrapper of ``tbb::flow::multifunction_node`` combined with
+concurrent hash table for storing the partial results. Let's also consider the ``multifunction_node`` to be ``tbb::flow::unlimited`` for simplicity.
+Since the elements from the input stream and the ``fold_stream_end`` indicator are submitted in the same manner, the current TBB implementation
+of ``multifunction_node`` would create and spawn _N + 1_ tasks, where _N_ is a number of elements in the input stream. Since the tasks are processed by
+the TBB scheduler in an unspecified order, it is possible that the task processing the ``fold_stream_end`` can be processed before other tasks processing
+the input stream.
+
+### ``fold_stream_end`` containing the number of elements in the stream
+
+The first implementation approach suggests ``fold_stream_end`` to contain the number of elements in the corresponding input stream.
+The ``fold_node`` should support an internal *signed* integer counter of elements for each input stream. Once the node is
+processing the element that is not a ``fold_stream_end``, the counter should be increased by 1 for each element. If the node is
+processing the ``fold_stream_end`` element, the counter should be decreased by _N_, where _N_ is a number of elements in the stream
+which is stored as part of the ``fold_stream_end``.
+
+In this case, having the internal counter not equal to _0_ indicates that there are still elements to process and once it is equal to _0_, the result
+of the fold is considered full and can be propagated to the ``fold_node`` successors.
```

Comment:
Making the single counter for both expected (negative count) and received (positive count) messages, which triggers at 0, has a downside - it loses the information about the number of messages and cannot pass it down. This value was computed as a byproduct of folding, and in theory it might not be known anywhere else, so it could possibly be useful in the subsequent processing.